Proposal: use Chordpro for PDF layout templates

martijnversluis opened the pull request.

isaiahdahl commented:
When running `yarn dev:pdf` a few odd things pop out

- [ ] somehting is double rendering text
![image](https://github.com/user-attachments/assets/24a2e2ab-9974-4258-a862-5f0466c68ca9)
- [ ] the conditional logic around pages don't seem to be working anymore. 

For example this was working:
```
"footer": {
            "height": 65,
            "content": [
                {
                    "type": "text",
                    "template": "%{copyright}",
                    "style": {
                        "name": "NimbusSansL-Reg",
                        "style": "normal",
                        "size": 7,
                        "color": "black",
                        "lineHeight": 1.2
                    },
                    "position": {
                        "x": "left",
                        "y": 20,
                        "width": 360
                    },
                    "condition": {
                        "and": [
                            {
                                "page": {
                                    "first": true
                                }
                            }
                        ]
                    }
                }
            ]
        }
```

Is this proposing using chordpro just for the "template" syntax or also to replace how the "condition" works.

isaiahdahl reviewed (commented):
When running `yarn dev:pdf` a few odd things pop out

- [ ] somehting is double rendering text
![image](https://github.com/user-attachments/assets/24a2e2ab-9974-4258-a862-5f0466c68ca9)
- [ ] the conditional logic around pages don't seem to be working anymore. 

For example this was working:
```
"footer": {
            "height": 65,
            "content": [
                {
                    "type": "text",
                    "template": "%{copyright}",
                    "style": {
                        "name": "NimbusSansL-Reg",
                        "style": "normal",
                        "size": 7,
                        "color": "black",
                        "lineHeight": 1.2
                    },
                    "position": {
                        "x": "left",
                        "y": 20,
                        "width": 360
                    },
                    "condition": {
                        "and": [
                            {
                                "page": {
                                    "first": true
                                }
                            }
                        ]
                    }
                }
            ]
        }
```

Is this proposing using chordpro just for the "template" syntax or also to replace how the "condition" works. 
With the default config in that dev:pdf command, the copyright line should only be on the first page

martijnversluis commented:
@isaiahdahl Thanks for checking! I resolved both issues and added tests to prevent regression.

- the double rendering was an actual double render call, leftover from a refactor I guess. The PDF formatter tests now checks the number of rendered items.
- the condition evaluation was broken due to `Metadata` only accepting `string`s. The metadata that is used evaluating conditionals is now a different type, preserving numbers.

The chordpro is just for evaluating the template. I introduced the `Condition` class which is responsible for evaluating the `condition`, it has its own unit tests to ensure its working.

isaiahdahl commented:
Actually it does look like this isn't working:
![image](https://github.com/user-attachments/assets/44694b84-9b1d-4edd-b708-5957355d6977)

isaiahdahl commented:
For that other PR I can write I was able to get things to a point where it looks like this:
![image](https://github.com/user-attachments/assets/7e3a142a-688d-4266-ade5-7a984c27b829)


and has a config that looks like this:
![image](https://github.com/user-attachments/assets/ed3a09f1-4ca0-4525-ba1c-59f3c0ebbb8d)

martijnversluis commented:
> For that other PR I can write I was able to get things to a point where it looks like this: 
> 
> and has a config that looks like this:

Wow that's really cool!

martijnversluis commented:
> Actually it does look like this isn't working: ![image](https://private-user-images.githubusercontent.com/23046374/416407700-44694b84-9b1d-4edd-b708-5957355d6977.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA1OTgyODksIm5iZiI6MTc0MDU5Nzk4OSwicGF0aCI6Ii8yMzA0NjM3NC80MTY0MDc3MDAtNDQ2OTRiODQtOWIxZC00ZWRkLWI3MDgtNTk1NzM1NWQ2OTc3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI2VDE5MjYyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRmOGViYjczMTVlMDZlZmY5NmI3NzRlY2I1MzgxYTljMmNlMGJmYTg2ZGQ0NTFhMjJhMjE3MzZhNGU4YWE2YmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.4bxrtocOf1jd3vMJwwLGkGA-i8We9kunJu8YwLD7pTI)

My bad, I renamed those to match [the official chordpro names](https://chordpro.org/chordpro/chordpro-configuration-format-strings/#page-related-metadata). `currentPage` became `page` and `totalPages` became `pages`. I fixed the config examples.

isaiahdahl commented:
In a previous version, it was working to not try and render the capo slot in the metadata when it didn't exist.

This is what it looks like now:
```
                    "template": "%{key|Key: %{}} · %{capo|Capo: %{capoKey}} · %{tempo|Tempo: %{}} · %{time|Time: %{}}",
```

![image](https://github.com/user-attachments/assets/a67bce92-b8f3-43e7-bd07-0e6158aac34e)

Notice how when no capo is set it shows the two dots back to back. And then when there's a capo it shows.

![image](https://github.com/user-attachments/assets/be6a74ea-9133-47e4-8a27-5137cda111c5)


How can we wrap the capo area including like one of those dots in conditional logic to not render the extra dots between. Ideally we actually want to do that for any of those meta's.

thoughts?

isaiahdahl commented:
Nvm I wasn't really thinking. 

Can easily be fixed with by changing the template to be 

```
%{key|Key: %{}} %{capo|· Capo: %{capoKey}} %{tempo|· Tempo: %{}} %{time|· Time: %{}}
```
